src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp

```diff
@@ -0,0 +1,564 @@
+                    return false;
+            }
+
+            if(ABlockTransferSrcVectorDim >= ABlockTransferThreadClusterArrangeOrder.size())
+                return false;
+
+            if(!(thread_slice_lengths[2] % BBlockTransferSrcScalarPerVector == 0))
+                return false;
+
+            if(!(thread_slice_lengths[2] % BBlockTransferDstScalarPerVector_K1 == 0))
+                return false;
+        }
+        // C Matrix
+        {
+            if(CThreadTransferSrcDstAccessOrder.size() <= CThreadTransferSrcDstVectorDim)
+                return false;
+        }
+        return true;
+    }
+};
+
+inline static auto generate_tunable_list_conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk()
+{
+    constexpr auto f32 = ck::DataTypeEnum_t::Float;
+    constexpr auto f16 = ck::DataTypeEnum_t::Half;
+
+    return std::vector<TunableConvIgemmFwdV4r4r4XdlopsNhwcKyxcNhwk>{
+        // clang-format off
+        // fp32
+        {f32, f32, 256, 256, 128, 4, 32, 32, 4, 4, 2, {1, 4, 4}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 4, 4, false, 
+            {1, 2, 4}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 4, 4, false, {2, 3, 0, 1, 7, 5, 4, 6}, 7, 1, 1, 1},
+        {f32, f32, 256, 128, 128, 4, 32, 32, 4, 2, 2, {1, 2, 4}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 4, 4, false, 
+            {1, 2, 4}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 4, 4, false, {2, 3, 0, 1, 7, 5, 4, 6}, 7, 1, 1, 1},
+
+        // fp16
+        {f16, f16, 256, 256, 256, 4, 32, 32, 8, 4, 4, {1, 4, 8}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, 
+            {1, 4, 8}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, {2, 3, 0, 1, 7, 5, 4, 6}, 7, 1, 1, 1},
+        {f16, f16, 256, 256, 128, 4, 32, 32, 8, 4, 2, {1, 4, 8}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, 
+            {1, 2, 8}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, {2, 3, 0, 1, 7, 5, 4, 6}, 7, 1, 1, 1},
+        {f16, f16, 256, 128, 256, 4, 32, 32, 8, 2, 4, {1, 2, 8}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, 
+            {1, 4, 8}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, {2, 3, 0, 1, 7, 5, 4, 6}, 7, 1, 1, 1},
+        {f16, f16, 256, 128, 128, 4, 32, 32, 8, 2, 2, {1, 2, 8}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, 
+            {1, 2, 8}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, {2, 3, 0, 1, 7, 5, 4, 6}, 7, 1, 1, 1},
+        {f16, f16, 256, 128, 64, 4, 32, 32, 8, 2, 2, {1, 4, 8}, {4, 32, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, 
+            {1, 2, 8}, {4, 32, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, {2, 3, 0, 1, 7, 5, 4, 6}, 7, 1, 1, 1},
+        {f16, f16, 256, 128, 64, 4, 32, 32, 8, 2, 1, {1, 2, 8}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, 
+            {1, 1, 8}, {4, 64, 1}, {1, 0, 2}, {1, 0, 2}, 2, 8, 8, false, {2, 3, 0, 1, 7, 5, 4, 6}, 7, 1, 1, 1},
+
+        // i8
+
+        // clang-format on
+    };
+}
+
+// TODO make this common interface and write specs for it
+struct ConvIgemmFwdV4r4r4XdlopsNhwcKyxcNhwk
```

Comment:
Please move definitions of all non-small methods (e.g. > 5 lines) to the corresponding .cpp file (filename should match header name).

Reply:
If there is no virtual function and no information hiding, is it necessary to separate?

Reply:
There are so many implementation details that they clutter up the interface and hinder readability. Perhaps this is a matter of taste, let's ask others.

CC @JehandadKhan @junliume

Reply:
I would agree with @atamazov  on the grounds that moving definitions to cpp files improves compilation and is better design.

Reply:
not resolved

Reply:
done